Replace gfs_cyc with an interval

WalterKolczynski-NOAA opened the pull request:
# Description
To facilitate longer and more flexible GFS cadences, the `gfs_cyc` variable is replaced with a specified interval. Up front, this is
reflected in a change in the arguments for setup_exp to:
```
--interval <n_hours>
```
Where `n_hours` is the interval (in hours) between gfs forecasts. `n_hours` must be a multiple of 6. If 0, no gfs will be run (only
gdas; only valid for cycled mode). The default value is 6 (every cycle). (This is a change from current behavior of 24.)

In cycled mode, there is an additional argument to control which cycle will be the first gfs cycle:
```
--sdate_gfs <YYYYMMDDHH>
```
The default if not provided is `--idate` + 6h (first full cycle). This is the same as current behavior when `gfs_cyc` is 6, but may vary from current behavior for other cadences.

As part of this change, some of the validation of the dates has been added. `--edate` has also been made optional and defaults to `--idate` if not provided.

During `config.base` template-filling, `INTERVAL_GFS` (renamed from `STEP_GFS`) is defined as `--interval` and `SDATE_GFS as
`--sdate_gfs`.

Some changes were necessary to the gfs verification (metp) job, as `gfs_cyc` was being used downstream by verif-global. That has been removed, and instead workflow will be responsible for only running metp on the correct cycles. This also removes "do nothing" metp tasks that exit immediately, because only the last GFS cycle in a day would actually process verification.

Now, metp has its own cycledef and will (a) always runs at 18z, regardless of whether gfs is running at 18z or not, if the interval is less than 24h; (b) use the same cycledef as gfs if the interval is 24h or greater. This is simpler than trying to determine the last gfs cycle of a day when it could change from day to day. To facilitate this change, support for the
undocumented rocoto dependency tag `taskvalid` is added, as the metp task needs to know whether the cycle has a gfsarch task or not. metp will trigger on gfsarch completing (as before), or look backwards for the last gfsarch to exist.

Additionally, a couple EE2 issues with the metp job are resolved (even though it is not run in ops):
- verif-global update replaced `$CDUMP` with `$RUN`
- `$DATAROOT` is no longer redefined in the metp job

Depends on NOAA-EMC/EMC_verif-global#137
Resolves #260
Refs #1299

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? **YES**
- Does this change require a documentation update? **YES**
- Does this change require an update to any of the following submodules? **YES**
  - [x] EMC verif-global (NOAA-EMC/EMC_verif-global#137)

# How has this been tested?
- Cycled test with 24-h interval on Hercules
- 

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

WalterKolczynski-NOAA committed:
Replace gfs_cyc with interval

To facilitate longer and more flexible GFS cadences, the `gfs_cyc`
variable is replaced with a specified interval. Up front, this is
reflected in a change in the arguments for setup_exp to:

```
--interval <n_hours>
```

Where `n_hours` is the interval (in hours) between gfs forecasts.
`n_hours` must be a multiple of 6. If 0, no gfs will be run (only
gdas; only valid for cycled mode). The default value is 6 (every
cycle).

In cycled mode, there is an additional argument to control which
cycle will be the first gfs cycle:

```
---sdate_gfs <YYYYMMDDHH>
```

The default if not provided is `--idate` + 6h (first full cycle).

As part of this change, some of the validation of the dates has
been added. `--edate` has also been made optional and defaults to
`--idate` if not provided.

During `config.base` template-filling, `INTERVAL_GFS` (renamed from
`STEP_GFS`) is defined as `--interval` and `SDATE_GFS as
`--sdate_gfs`.

Some changes were necessary to the gfs verification (metp) job, as
`gfs_cyc` was being used downstream by verif-global. That has been
removed, and instead workflow will be responsible for only running
metp on the correct cycles. This also removes "do nothing" metp
tasks that exit immediately, because only the last GFS cycle in a
day would actually process verification.

Now, metp has its own cycledef and always runs at 18z, regardless
of whether gfs is running at 18z or not. This is simplier than
trying to determine the last gfs cycle of a day when it could
change from day to day. To facilitate this change, support for the
undocumented rocoto dependency tag `taskvalid` is added, as the
metp task needs to know whether the cycle has a gfsarch task or not.
metp will trigger on gfsarch completing (as before), or gdasarch
completing if there is no gfsarch.

metp tasks are no longer generated for forecast-only, as the pgbanl
files (copied of the 1p00 pgbanl files) are not generated for f-o
anyway. If metp is needed for f-o, additional work will be needed.

Additionally, a couple EE2 issues with the metp job are resolved
(even though it is not run in ops):
- verif-global update replaced `$CDUMP` with `$RUN`
- `$DATAROOT` is no longer redefined in the metp job

Depends on NOAA-EMC/EMC_verif-global#137
Resolves #260
Refs #1299

WalterKolczynski-NOAA committed:
Update CI tests for gfs_cyc to interval change

WalterKolczynski-NOAA committed:
Update documentation for change to interval from gfs_cyc

WalterKolczynski-NOAA committed:
Fix overindented line in setup_expt

WalterKolczynski-NOAA commented:
More manual testing to come, but ready enough to get eyes on it.

WalterKolczynski-NOAA committed:
Revert inadvertent change to C384C192_hybatmda test

WalterKolczynski-NOAA committed:
Remove unnecessary declaration

WalterKolczynski-NOAA committed:
Revert inadvertent cycledef change for prep and atmanlinit

WalterKolczynski-NOAA committed:
Fix interval issue with GEFS XML generation

WalterKolczynski-NOAA committed:
Bugfixes for verif-global

KateFriedman-NOAA commented on workflow/applications/gfs_cycled.py:
```diff
@@ -128,7 +128,7 @@ def _get_app_configs(self):
     @staticmethod
     def _update_base(base_in):
 
-        return GFSCycledAppConfig.get_gfs_cyc_dates(base_in)
+        return base_in
```
Is this function needed anymore? Doesn't look like it's doing anything anymore and just returns what goes into it.

DavidHuber-NOAA replied:
It gets called by the superclass (`AppConfig._init_finalize`), so it does need to be defined.

KateFriedman-NOAA replied:
Ah, thanks for explaining @DavidHuber-NOAA !

WalterKolczynski-NOAA committed:
Correct metp cycledefs

To avoid running metp on days where there is no gfs, the cycledefs
are adjusted somewhat. First, if the interval is >= 24h, the metp
cycledef will be identical to gfs. If the interval is < 24h, it
remains 18z every day (except the last). Second, a last_gfs is
added so metp will run on for the last gfs cycle even if it there
is no gdas cycle for 18z that day. This required computing the real
gfs end date to use as the last cycle.

WalterKolczynski-NOAA committed:
Fix missing whitespace around operator

WalterKolczynski-NOAA committed:
Fix cleanup dependency when metp is not run that cycle

WalterKolczynski-NOAA commented:
Ready for review. Also encourage third-party testing in excess of what CI will cover due to the nature of the change. I ran some different multi-day experiments myself, but outside validation would be appreciated.

WalterKolczynski-NOAA committed:
Fix cleanup dependency for forecast-only

DavidHuber-NOAA commented:
I just tested the C48_ATM case at the head of develop (https://github.com/NOAA-EMC/global-workflow/commit/8f0541cd61755e74e1f3116dedcc3afc8fa9cda1) on Orion and the METplus output statistics are generated for the `metpg2g1` jobs.  `Grid2Grid` statistics are generated based on analyses by default, but in forecast-only mode, the code fails back to the 0-hour forecast (see outputs in `/work/noaa/global/dhuber/archive/metplus_data/by_VSDB/grid2grid/{anom,pres,sfc}/12Z/c48_atm`).

The `metppcp1` job does not produce data, and this is expected, as it requires at least two cycles at least 24 hours apart to produce statistics.  However, if a user runs subsequent cycles in forecast-only mode, then I would expect this job to produce statistics.

The `metpg2o1` job does not produce data because the NAM `prepbufr` input files are missing for the requisite dates (2021032312 and 2021032318).  However, if these are uploaded to Orion, I expect this job to run to completion as well as it does not need an analysis file to perform the statistics (only f00 forecasts and NAM are used as baselines).

As such, I think that the `metp` jobs should stay enabled in forecast-only mode.

TerrenceMcGuinness-NOAA commented:
Jenkins Controller has this PR stalled without an apparent error being reported.  I am going to force quit it and restart the PR CI from scratch.

DavidHuber-NOAA commented:
Apologies, my analysis of the `metpg2o1` job was not correct.  The job failed because of duplicate attempts to create a `prepbufr` directory, indicating that the 'fix' in #2907 was not 100% successful.  I will try rerunning and see if it is successful on the next run.

EDIT: nope, still failed.

DavidHuber-NOAA commented:
Increasing the sleep time to 5s/job was successful on Orion for the `C48_ATM` `gfsmetpg2o1` job.  However, additional testing is required to verify that this would work consistently.  Perhaps for now the metp jobs should return to running serially.

WalterKolczynski-NOAA commented:
> Increasing the sleep time to 5s/job was successful on Orion for the `C48_ATM` `gfsmetpg2o1` job. However, additional testing is required to verify that this would work consistently. Perhaps for now the metp jobs should return to running serially.

It was never actually changed back to parallel: https://github.com/NOAA-EMC/global-workflow/blob/develop/workflow/rocoto/gfs_tasks.py#L1869

DavidHuber-NOAA commented:
Hmm, that's interesting, because it is launching an MPMD job:
https://github.com/NOAA-EMC/EMC_verif-global/blob/e7e6bc43584e0b8911819b8f875cc8ee747db76d/ush/create_METplus_job_scripts.py#L1897-L1959
since `MPMD` is set to `YES` and `nproc` is set to `4` in verif-global's [run_verif_global_in_global_workflow.sh script](https://github.com/NOAA-EMC/EMC_verif-global/blob/e7e6bc43584e0b8911819b8f875cc8ee747db76d/ush/run_verif_global_in_global_workflow.sh#L278-L279) and the global workflow in [config.resources](https://github.com/NOAA-EMC/global-workflow/blob/8f0541cd61755e74e1f3116dedcc3afc8fa9cda1/parm/config/gfs/config.resources#L992) and later in [config.metp](https://github.com/NOAA-EMC/global-workflow/blob/8f0541cd61755e74e1f3116dedcc3afc8fa9cda1/parm/config/gfs/config.metp#L11).

Also, I think the is_serial flag doesn't specify the job itself as being serial, but rather that the jobs in the metatask are to be run sequentially, i.e. `gfsmetppcp1` depends on `gfsmetpg2o1` depends on `gfsmetpg2g1`.  This is documented in the [rocoto documentation](https://christopherwharrop.github.io/rocoto/) as the last item in the docs.  I don't think this is the desired configuration, though.  They should be allowed to run simultaneously.

As an example, `/work2/noaa/global/dhuber/para/hercules/EXPDIR/test_cp/test_cp.xml` shows
```xml
<metatask name="gfsmetp" mode="serial">

    <var name="metpcase">g2g1 g2o1 pcp1</var>

    <task name="gfsmetp#metpcase#" cycledefs="gfs" maxtries="&MAXTRIES;">

...
        <walltime>06:00:00</walltime>
        <nodes>1:ppn=4:tpp=1</nodes>
```
with `ppn=4`.

WalterKolczynski-NOAA commented:
> Also, I think the is_serial flag doesn't specify the job itself as being serial, but rather that the jobs in the metatask are to be run sequentially, i.e. `gfsmetppcp1` depends on `gfsmetpg2o1` depends on `gfsmetpg2g1`. This is documented in the [rocoto documentation](https://christopherwharrop.github.io/rocoto/) as the last item in the docs. I don't think this is the desired configuration, though. They should be allowed to run simultaneously.
> 

Yes, you are correct. I was thinking of the tasks themselves, not the MPMD.

WalterKolczynski-NOAA committed:
Re-enable metp for forecast-only

Turns metp back on for forecast-only mode. The metp cycledef follows
the same practice of cycled, where if the interval is under 24, it
will run every 18z. But if the interval is 24+, the cycledef will
follow the gfs cycledef.

Since there is no gdasarch task to key off of when gfsarch does not
exist for the 18z cycle, the dependency is now more complicated and
progressively works backwards to find a cycle that exists, then uses
the gfsarch for that cycle.

WalterKolczynski-NOAA committed:
Make metp tasks parallel again

WalterKolczynski-NOAA commented:
@DavidHuber-NOAA I've restored metp for f-o and made the tasks parallel again.

WalterKolczynski-NOAA committed:
Fix pynorms

WalterKolczynski-NOAA committed:
Fix metp forecast-only cycledef

DavidHuber-NOAA commented:
> > Also, I think the is_serial flag doesn't specify the job itself as being serial, but rather that the jobs in the metatask are to be run sequentially, i.e. `gfsmetppcp1` depends on `gfsmetpg2o1` depends on `gfsmetpg2g1`. This is documented in the [rocoto documentation](https://christopherwharrop.github.io/rocoto/) as the last item in the docs. I don't think this is the desired configuration, though. They should be allowed to run simultaneously.
> 
> Yes, you are correct. I was thinking of the tasks themselves, not the MPMD.

Aha, understood.

WalterKolczynski-NOAA commented:
I think I need @malloryprow to look at these logs. I'm having trouble determining which "errors" are expected and which need to be fixed.

malloryprow commented:
@WalterKolczynski-NOAA Can you point me to the log(s) you want me to look at?

KateFriedman-NOAA commented:
I cloned, built, and linked the `feature/interval` branch on WCOSS2. I ran setup for a 5-day cycled gfs experiment multiple times using different intervals `[0, 1, 6, 12, 18, 24, 120]` and providing `--sdate_gfs` or not. It correctly errored with a message when I tried `--interval 1`. The resulting xmls included the gfs suite in the expected cycles for each tested interval. I did not run any of the experiments, so I can not comment on the metp aspect.

KateFriedman-NOAA reviewed (commented):
Changes look good. No comments or requested changes from me right now. I tested the xml generation with various intervals and they looked good. See my https://github.com/NOAA-EMC/global-workflow/pull/2928#issuecomment-2391966476. Will return to rereview and approve once metp issues are resolved.

DavidHuber-NOAA commented:
@malloryprow Here is a sample log on Hera: `/scratch1/NCEPDEV/global/CI/2928/RUNTESTS/COMROOT/C48_ATM_0c600afa/logs/2021032318/gfsmetpg2o1.log`

malloryprow commented:
Something is up with the valid times it is gathering for. It is only getting valid times for 20210323 12Z-21Z. and nothing 00Z-09Z.

WalterKolczynski-NOAA commented:
> Something is up with the valid times it is gathering for. It is only getting valid times for 20210323 12Z-21Z. and nothing 00Z-09Z.

This is for a 12z-only forecast (except metp, which runs at 18z).

malloryprow commented:
Gotcha, well something isn't right. For grid2obs it can still verify all valid hours regardless if it is just a 12Z forecast.

malloryprow commented:
If that is not the goal, then it is still trying to verify forecasts valid 00Z-09Z and that is why the ERROR missing file messages are happening.

WalterKolczynski-NOAA commented:
> If that is not the goal, then it is still trying to verify forecasts valid 00Z-09Z and that is why the ERROR missing file messages are happening.

It's fine to verify 00z valid times, but the experiment is only one cycle, so it would have to be using something outside of workflow to do that. It seemed like it was using obs you had staged somewhere for the verification.

malloryprow commented:
> > If that is not the goal, then it is still trying to verify forecasts valid 00Z-09Z and that is why the ERROR missing file messages are happening.
> 
> It's fine to verify 00z valid times, but the experiment is only one cycle, so it would have to be using something outside of workflow to do that. It seemed like it was using obs you had staged somewhere for the verification.

Yes! It is using prepbufr files from the archive I keep.

DavidHuber-NOAA commented:
I ran a develop version of C48_ATM and compared logs.  It looks like the problems start at `/scratch1/NCEPDEV/global/CI/2928/RUNTESTS/COMROOT/C48_ATM_0c600afa/logs/2021032318/gfsmetpg2o1.log` line 2159, when attempting to read the NAM prepbufr file for 2021032318:
```
10/03 04:50:48.762 metplus.PB2NC (command_builder.py:122) ERROR: (command_builder.py:539) Could not find OBS_INPUT file /scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C48_ATM_0c600afa/gfs.2021032318/metpg2o1.398127/grid2obs_step1/data/prepbufr/prepbufr.nam.2021032318 using template prepbufr.nam.{valid?fmt=%Y%m%d%H}
```
This file is found correctly in develop.  For comparison, see `/scratch1/NCEPDEV/global/David.Huber/para/baseline_metp/COMROOT/c48_atm/logs/2021032312/gfsmetpg2o1.log`.

Earlier in the `C48_ATM_0c600afa` `gfsmetpg2o1.log`, a message is written that this file is linked correctly.  Just guessing here, but I suspect that the template naming convention reported at line 2159 is using a different hour than reported, i.e. `{valid?fmt=%Y%m%d%H}`.  It's either 1) just that, 2) the link is not actually being created or at some point is deleted, or 3) the reported link differs from what is actually created.

WalterKolczynski-NOAA commented:
> I ran a develop version of C48_ATM and compared logs. It looks like the problems start at `/scratch1/NCEPDEV/global/CI/2928/RUNTESTS/COMROOT/C48_ATM_0c600afa/logs/2021032318/gfsmetpg2o1.log` line 2159, when attempting to read the NAM prepbufr file for 2021032318:
> 
> ```
> 10/03 04:50:48.762 metplus.PB2NC (command_builder.py:122) ERROR: (command_builder.py:539) Could not find OBS_INPUT file /scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C48_ATM_0c600afa/gfs.2021032318/metpg2o1.398127/grid2obs_step1/data/prepbufr/prepbufr.nam.2021032318 using template prepbufr.nam.{valid?fmt=%Y%m%d%H}
> ```
> 
> This file is found correctly in develop. For comparison, see `/scratch1/NCEPDEV/global/David.Huber/para/baseline_metp/COMROOT/c48_atm/logs/2021032312/gfsmetpg2o1.log`.
> 
> Earlier in the `C48_ATM_0c600afa` `gfsmetpg2o1.log`, a message is written that this file is linked correctly. Just guessing here, but I suspect that the template naming convention reported at line 2159 is using a different hour than reported, i.e. `{valid?fmt=%Y%m%d%H}`. It's either 1) just that, 2) the link is not actually being created or at some point is deleted, or 3) the reported link differs from what is actually created.

Hmmm. I didn't make any changes to that. I wonder if the issue was something ephemeral unrelated to this PR.

DavidHuber-NOAA commented:
@malloryprow Could you take a look at the log files mentioned [above](https://github.com/NOAA-EMC/global-workflow/pull/2928#issuecomment-2394349063)?  Do you think this is a misreporting by one of the scripts?

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/interval

WalterKolczynski-NOAA committed:
Update new replay case for interval

malloryprow commented:
Hi sorry, I've been on sick leave. Not sure if anyone has investigated the logs in my absence. Do you still need me to look?

TerrenceMcGuinness-NOAA commented:
```
Submodule 'ccpp' (https://github.com/ufs-community/ccpp-physics.git) registered for path 'sorc/ufs_utils.fd/ccpp-physics'
Cloning into '/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2928/global-workflow/sorc/ufs_utils.fd/ccpp-physics'...
Submodule path 'sorc/ufs_utils.fd/ccpp-physics': checked out '89ddce75d4d64252f1871c095a362ad391065b89'
Submodule 'physics/Radiation/RRTMGP/rte-rrtmgp' (https://github.com/earth-system-radiation/rte-rrtmgp) registered for path 'sorc/ufs_utils.fd/ccpp-physics/physics/Radiation/RRTMGP/rte-rrtmgp'
Cloning into '/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2928/global-workflow/sorc/ufs_utils.fd/ccpp-physics/physics/Radiation/RRTMGP/rte-rrtmgp'...
Submodule path 'sorc/ufs_utils.fd/ccpp-physics/physics/Radiation/RRTMGP/rte-rrtmgp': checked out '74a0e098b2163425e4b5466c2dfcf8ae26d560a5'
fatal: remote error: upload-pack: not our ref b2ee80cac7921a3016fa5a857cc58acfccc4baea
fatal: Fetched in submodule path 'sorc/verif-global.fd', but it did not contain b2ee80cac7921a3016fa5a857cc58acfccc4baea. Direct fetching of that commit failed.
fatal: 
failed to run git: exit status 128
```

DavidHuber-NOAA commented:
@malloryprow Yes, if you could take a look at the logs, that would be very helpful.  Thanks!

malloryprow commented:
Can you post the paths for me again? Sorry!

DavidHuber-NOAA commented:
Sure, you can find them [here](https://github.com/NOAA-EMC/global-workflow/pull/2928#issuecomment-2394349063).  I just checked and all of the logs are all still available.

malloryprow commented:
Could the directory tried to have been scrubbed during the run or something? I don't think the code would have reported that and not linked it. Have we tried rerunning?

malloryprow commented:
And if re-run can we do KEEPDATA=YES?

DavidHuber-NOAA commented:
I can rerun the develop branch that way.  @WalterKolczynski-NOAA would you be able to rerun the `C48_ATM` test case for your branch with `KEEPDATA=YES`?

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/interval

WalterKolczynski-NOAA commented:
@malloryprow Possibly related, it looks like NOAA-EMC/EMC_verif-global#137 never got merged. I thought for sure it had. I'm not sure how this worked at all in CI (WCOSS caught it). That should be merged before we run any more CI.

malloryprow commented:
Oh sorry, I'll merge it.

WalterKolczynski-NOAA commented:
@malloryprow Thanks. I've got a test case running on Hercules now. If that doesn't show the problem seen in CI, I'm going to assume that was the problem. If it does, KEEPDATA is on so I'll have an example for you.

WalterKolczynski-NOAA committed:
Update metp dependencies for new task names

WalterKolczynski-NOAA commented:
Once I fixed the issue, this runs for me on Hercules. Going to try CI again and see if this is still an issue.

WalterKolczynski-NOAA committed:
Fix extract vars dependency

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA The gfscleanup job is running before the gfs_metpg2o1 job is completing:

```
202112210000            gfs_metpg2g1                     2777753           SUCCEEDED                   0         2          89.0
202112210000            gfs_metpg2o1                     2777754                DEAD                   1         2          98.0
202112210000            gfs_metppcp1                     2777744           SUCCEEDED                   0         1          27.0
202112210000                gfs_arch                     2777704           SUCCEEDED                   0         1          84.0
202112210000             gfs_cleanup                     2777745           SUCCEEDED                   0         1          28.0
```

DavidHuber-NOAA commented:
The reason is that the dependency isn't quite right.  The `taskvalid` check is looking for the `task` `gfs_metp`, but `gfs_metp` is a `metatask`.  So this

```xml
    <dependency>
        <and>
            <taskdep task="gfs_arch"/>
            <or>
                <not><taskvalid task="gfs_metp"/></not>
                <metataskdep metatask="gfs_metp"/>
            </or>
        </and>
    </dependency>
```

Should be
```xml
    <dependency>
        <and>
            <taskdep task="gfs_arch"/>
            <or>
                <not><taskvalid metatask="gfs_metp"/></not>
                <metataskdep metatask="gfs_metp"/>
            </or>
        </and>
    </dependency>
```

However, I'm not sure if `taskvalid` supports checking `metatasks`.  If not, then it should check if any of the `gfs_metp*` (`g2g1`, `g2o1`, or `pcp1`) tasks are valid.

WalterKolczynski-NOAA commented:
> The reason is that the dependency isn't quite right. The `taskvalid` check is looking for the `task` `gfs_metp`, but `gfs_metp` is a `metatask`. So this
> 
> ```
>     <dependency>
>         <and>
>             <taskdep task="gfs_arch"/>
>             <or>
>                 <not><taskvalid task="gfs_metp"/></not>
>                 <metataskdep metatask="gfs_metp"/>
>             </or>
>         </and>
>     </dependency>
> ```
> 
> Should be
> 
> ```
>     <dependency>
>         <and>
>             <taskdep task="gfs_arch"/>
>             <or>
>                 <not><taskvalid metatask="gfs_metp"/></not>
>                 <metataskdep metatask="gfs_metp"/>
>             </or>
>         </and>
>     </dependency>
> ```
> 
> However, I'm not sure if `taskvalid` supports checking `metatasks`. If not, then it should check if any of the `gfs_metp*` (`g2g1`, `g2o1`, or `pcp1`) tasks are valid.

That doesn't explain all of them. For C48_ATM, metp is alone in its own cycle, so there shouldn't be any premature cleanup issue there.

WalterKolczynski-NOAA committed:
Fix gfs_cleanup dependency

While `taskvalid` does not complain about using a metatask name, using
one will always return as not valid.

WalterKolczynski-NOAA committed:
Make metp serial again

metp is using non-standard DATA directories and the same for all
tasks, so when one task remove the directory, that interferes with
other jobs using it.

DavidHuber-NOAA commented:
The gfs_metpg2o1 job failed on Hercules due to duplicate attempts to make the same directory (#2971).  I think we will need to revert
https://github.com/NOAA-EMC/global-workflow/blob/ac3cde510d4aeed632e3b080b6227cb5a26ef6d0/parm/config/gfs/config.resources#L991-L992
to
```bash
    ntasks=1
    tasks_per_node=1
```

DavidHuber-NOAA commented:
The C48_ATM case on Hera failed the post jobs because the inline post did not create master grib2 files for hours 0-60.  Similarly, the netCDF outputs were not created for hours 0-60.  Looking through the logs, it appears that the restart files were found in `/scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C48_ATM_c9e3174a/gfs.2021032312/gfsfcst.2021032312/restart/FV3_RESTART`.  So a fresh commit needs to be made or the `DATAROOT` directory needs to be cleared out.

DavidHuber-NOAA committed:
Run the metp jobs with one task each

WalterKolczynski-NOAA committed:
Merge pull request #27 from DavidHuber-NOAA/fix/single_metp_task

Run the metp jobs with one task each

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/interval

DavidHuber-NOAA commented on workflow/rocoto/rocoto.py:
```diff
@@ -296,6 +297,27 @@ def _add_cycle_tag(dep_dict: Dict[str, Any]) -> str:
+    :rtype: str
+    """
+
+    dep_type = dep_dict.get('type', None)
+    dep_name = dep_dict.get('name', None)
+
+    if dep_name is None:
+        msg = f'a {dep_type} name is necessary for {dep_type} dependency'
+        raise KeyError(msg)
+
```
Should a similar check be made for a `None` `dep_type`?

WalterKolczynski-NOAA replied:
If the type is `None`, it never reaches this method (an exception will be thrown on L196).

DavidHuber-NOAA commented on workflow/setup_expt.py:
```diff
@@ -186,6 +187,19 @@ def input_args(*argv):
     ufs_apps = ['ATM', 'ATMA', 'ATMW', 'S2S', 'S2SA', 'S2SW', 'S2SWA']
 
+    def _validate_interval(interval_str):
+        err_msg = f'must be a non-negative integer multiple of 6 ({interval_str} given)'
```
Should it always be a multiple of 6 or should it be a multiple of `assim_freq` for cycled runs?  I'm not sure if anyone changes `assim_freq`, so it may be a moot question.

WalterKolczynski-NOAA replied:
`assim_freq` (see comment below in the code), but I don't think there is a way to know it at this point, so I had to hard-code it.

DavidHuber-NOAA reviewed (dismissed):
Looks good.  Approve pending successful WCOSS2 testing.

WalterKolczynski-NOAA committed:
Fix cycled metp cycledef

TerrenceMcGuinness-NOAA commented:
Starting over from scratch with Walter's **`d54bc65`** update.

WalterKolczynski-NOAA commented:
Sync'd missing snow observation data to cactus.

WalterKolczynski-NOAA committed:
Turn off gefs replay test on WCOSS

The UFS model executable is not built for the structured wave grid
used by GEFS on WCOSS, so GEFS tests should not run on WCOSS.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/interval

WalterKolczynski-NOAA commented:
Failed because restarts from previous attempt weren't wiped, so forecast picked up in the middle.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/interval

WalterKolczynski-NOAA commented:
All manual CI tests passed on WCOSS:
```
08:05:32 EXPDIR/>while [[ 1 ]]; do echo; echo; date; for d in C*; do echo "******** ${d} ********"; cd $d; rocotostat -d ${d}.db -w ${d}.xml | tail -n 1; echo; cd ..; done; echo; sleep 300; done


Tue Oct 22 08:05:39 UTC 2024
******** C48_ATM ********
202103231800            gfs_metppcp1                   158916037           SUCCEEDED                   0         1          15.0

******** C48_S2SW ********
202103231800            gfs_metppcp1                   158916989           SUCCEEDED                   0         1          16.0

******** C96_atm3DVar_extended ********
202112211800             gfs_cleanup                   158951431           SUCCEEDED                   0         1          17.0

******** C96C48_hybatmaerosnowDA ********
202112210000        enkfgdas_cleanup                   158920982           SUCCEEDED                   0         1          15.0

******** C96C48_hybatmDA ********
202112210600        enkfgdas_cleanup                   158921341           SUCCEEDED                   0         1          16.0

******** C96C48_ufs_hybatmDA ********
202402240600        enkfgdas_cleanup                   158925325           SUCCEEDED                   0         1          15.0
```

DavidHuber-NOAA reviewed (approved):
Thanks for all the work on this @WalterKolczynski-NOAA!

KateFriedman-NOAA reviewed (approved):
Looks good, thanks @WalterKolczynski-NOAA !

DavidHuber-NOAA commented on workflow/rocoto/gfs_forecast_only_xml.py:
```diff
@@ -12,15 +12,36 @@ def __init__(self, app_config: AppConfig, rocoto_config: Dict) -> None:
+
+        date2 = sdate_gfs + interval_gfs
+        if date2 <= edate_gfs:
+            date2_gfs_str = date2_gfs.strftime("%Y%m%d%H%M")
```
Just noticed that `date2_gfs` should be `date2`.  I'll fix this in #2943.

DavidHuber-NOAA replied:
I'm surprised PEP8 didn't complain about this.
